integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+        // Doing XAER_RMFAIL will put us in FINISH_ERROR.
+        this.computeAssociation(XARoutine.ROLLBACK,
+                                xid,
+                                EnumSet.of(Association.BranchState.IDLE,
+                                           Association.BranchState.PREPARED,
+                                           Association.BranchState.ROLLBACK_ONLY),
+                                LocalXAResource::rollbackAndReset,
+                                true); // do remove association on error
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "rollback");
+        }
+    }
+
+    @Override // XAResource
+    public void forget(Xid xid) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "forget", xid);
+        }
+        requireNonNullXid(xid);
+
+        this.computeAssociation(XARoutine.FORGET,
+                                xid,
+                                EnumSet.of(Association.BranchState.HEURISTICALLY_COMPLETED),
+                                LocalXAResource::forgetAndReset,
+                                false); // don't remove association on error
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "forget");
+        }
+    }
+
+    @Override // XAResource
+    public Xid[] recover(int flags) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "recover", flagsToString(flags));
+        }
+
+        switch (flags) {
+        case TMENDRSCAN:
+        case TMNOFLAGS:
+        case TMSTARTRSCAN:
+            break;
+        default:
+            // Bad flags.
+            throw (XAException) new XAException(XAER_INVAL)
+                .initCause(new IllegalArgumentException("flags: " + flagsToString(flags)));
+        }
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "recover", EMPTY_XID_ARRAY);
+        }
+        return EMPTY_XID_ARRAY;
+    }
+
+    @Override // XAResource
+    public boolean isSameRM(XAResource xaResource) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "isSameRM", xaResource);
+            LOGGER.exiting(this.getClass().getName(), "isSameRM", this == xaResource);
+        }
+        return this == xaResource;
+    }
+
+    @Override // XAResource
+    public int getTransactionTimeout() {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "getTransactionTimeout");
+            LOGGER.exiting(this.getClass().getName(), "getTransactionTimeout", 0);
+        }
+        return 0;
+    }
+
+    @Override // XAResource
+    public boolean setTransactionTimeout(int transactionTimeoutInSeconds) {
```

Comment:
Interesting point of note: this is the first method that is called by the transaction manager (at least in Narayana's implementation), even before `start`.